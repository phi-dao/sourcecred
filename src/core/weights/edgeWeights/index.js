// @flow

import {type EdgeWeightsI, EdgeWeights} from "./edgeWeights";

export {EdgeWeights};
export type {EdgeWeightsI};

import {
  type EdgeWeight,
  type EdgeOperator,
  type EdgeWeightsT,
  empty,
  copy,
  merge,
} from "./edgeWeightsT";

export type {EdgeWeight, EdgeOperator, EdgeWeightsT};

export {empty, copy, merge};
