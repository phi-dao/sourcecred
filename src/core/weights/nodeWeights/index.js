// @flow

import {type NodeWeightsI, NodeWeights} from "./nodeWeights";

export {NodeWeights};
export type {NodeWeightsI};

import {
  type NodeWeight,
  type NodeOperator,
  type NodeWeightsT,
  empty,
  copy,
  merge,
} from "./nodeWeightsT";

export type {NodeWeight, NodeOperator, NodeWeightsT};

export {empty, copy, merge};
