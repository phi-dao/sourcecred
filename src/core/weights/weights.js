// @flow

import {type NodeAddressT, type EdgeAddressT} from "../graph";
import {
  type NodeWeight,
  type NodeWeightsI,
  type NodeOperator,
  type NodeWeightsT,
  NodeWeights,
  empty as emptyNodeWeightsT,
} from "./nodeWeights";
import {
  type EdgeWeight,
  type EdgeWeightsI,
  type EdgeOperator,
  type EdgeWeightsT,
  EdgeWeights,
  empty as emptyEdgeWeightsT,
} from "./edgeWeights";
import {
  type WeightsT,
  type WeightsJSON,
  type WeightsTResolvers,
  compareWeightsT,
  toJSON,
  fromJSON as JsonToWeightsT,
  toWeightsT,
} from "./weightsT";

/**
 * WeightsI is an interface that WeightsT data
 * gets lifted into that provides helper functions.
 */
export type WeightsI = {
  eject: () => WeightsT,
  nodeWeightsT: () => NodeWeightsT,
  edgeWeightsT: () => EdgeWeightsT,
  getNodeWeight: (NodeAddressT) => NodeWeight,
  getEdgeWeight: (EdgeAddressT) => EdgeWeight,
  setNodeWeight: (NodeAddressT, NodeWeight) => WeightsI,
  setEdgeWeight: (EdgeAddressT, EdgeWeight) => WeightsI,
  merge: (
    ws: $ReadOnlyArray<WeightsI>,
    resolvers?: WeightsTResolvers
  ) => WeightsI,
  copy: () => WeightsI,
  toJSON: () => WeightsJSON,
};

/**
 * Weights composes NodeWeightsI and EdgeWeightsI into a higher
 * order object, allowing us to interface with the underlying
 * Node/EdgeWeights while holding the aggregated Weights here.
 */
export const Weights = (
  nodeWeights: NodeWeightsI = NodeWeights(),
  edgeWeights: EdgeWeightsI = EdgeWeights()
): WeightsI => {
  const {getNodeWeight, setNodeWeight, eject: nodeWeightsT} = nodeWeights;
  const {getEdgeWeight, setEdgeWeight, eject: edgeWeightsT} = edgeWeights;
  const eject = () => toWeightsT(nodeWeights.eject(), edgeWeights.eject());
  return {
    eject,
    nodeWeightsT,
    edgeWeightsT,
    getNodeWeight,
    getEdgeWeight,
    setNodeWeight: (address: NodeAddressT, weight: NodeWeight) =>
      Weights(setNodeWeight(address, weight), edgeWeights),
    setEdgeWeight: (address: EdgeAddressT, weight: EdgeWeight) =>
      Weights(nodeWeights, setEdgeWeight(address, weight)),
    merge: (
      ws: $ReadOnlyArray<WeightsI>,
      resolvers?: {|
        +nodeResolver: NodeOperator,
        +edgeResolver: EdgeOperator,
      |}
    ) => merge([Weights(nodeWeights, edgeWeights), ...ws], resolvers),
    copy: () => Weights(nodeWeights.copy(), edgeWeights.copy()),
    toJSON: () => toJSON(eject()),
  };
};

/**
 * Lifts Node/EdgeWeightsT instead of Node/EdgeWeights by abstracting
 * the intermediate transformatino to Node/EdgeWeights.
 */
export function weightsTToWeights(
  nodeWeightsT: NodeWeightsT = emptyNodeWeightsT(),
  edgeWeightsT: EdgeWeightsT = emptyEdgeWeightsT()
): WeightsI {
  return Weights(NodeWeights(nodeWeightsT), EdgeWeights(edgeWeightsT));
}

/**
 * Merge composes the merge methods on NodeWeights and EdgeWeights
 * into a higher order Weights level merge.
 */
function merge(
  ws: $ReadOnlyArray<WeightsI>,
  resolvers?: WeightsTResolvers
): WeightsI {
  const nodeResolver = resolvers ? resolvers.nodeResolver : undefined;
  const edgeResolver = resolvers ? resolvers.edgeResolver : undefined;
  const nodeWeightsT = NodeWeights();
  const edgeWeightsT = EdgeWeights();
  const mergedNodeWeights = nodeWeightsT.merge(
    ws.map((w) => NodeWeights(w.nodeWeightsT())),
    nodeResolver
  );
  const mergedEdgeWeights = edgeWeightsT.merge(
    ws.map((w) => EdgeWeights(w.edgeWeightsT())),
    edgeResolver
  );
  return Weights(mergedNodeWeights, mergedEdgeWeights);
}

export function fromJSON(json: WeightsJSON): WeightsI {
  const {nodeWeightsT, edgeWeightsT} = JsonToWeightsT(json);
  return weightsTToWeights(nodeWeightsT, edgeWeightsT);
}

/**
 * Predicate determines whether two WeightsI's are equivalent
 * checking the following conditions which are not handled by typing:
 * - Node/Edge weights are equivalent
 * - Functions match
 */
export function areEqual(W1: WeightsI, W2: WeightsI): boolean {
  const {weightsAreEqual} = compareWeightsT(W1.eject(), W2.eject());
  return (
    weightsAreEqual &&
    W1.copy.toString() === W2.copy.toString() &&
    W1.toJSON.toString() === W2.toJSON.toString() &&
    W1.merge.toString() === W2.merge.toString() &&
    W1.getNodeWeight.toString() === W2.getNodeWeight.toString() &&
    W1.getEdgeWeight.toString() === W2.getEdgeWeight.toString() &&
    W1.setNodeWeight.toString() === W2.setNodeWeight.toString() &&
    W1.setEdgeWeight.toString() === W2.setEdgeWeight.toString()
  );
}
