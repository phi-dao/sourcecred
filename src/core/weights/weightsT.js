// @flow

import deepEqual from "lodash.isequal";
import {
  type NodeAddressT,
  type EdgeAddressT,
  NodeAddress,
  EdgeAddress,
} from "../graph";
import * as C from "../../util/combo";
import {toCompat, type Compatible, compatibleParser} from "../../util/compat";
import * as MapUtil from "../../util/map";
import {
  type NodeWeight,
  type NodeOperator,
  type NodeWeightsT,
  empty as emptyNodeWeightsT,
  copy as copyNodeWeightsT,
} from "./nodeWeights";
import {
  type EdgeWeight,
  type EdgeOperator,
  type EdgeWeightsT,
  empty as emptyEdgeWeightsT,
  copy as copyEdgeWeightsT,
} from "./edgeWeights";

export type WeightsTResolvers = {|
  +nodeResolver: NodeOperator,
  +edgeResolver: EdgeOperator,
|};

/**
 * Represents the weights for nodes and edges.
 *
 * The weights are stored by address prefix, i.e. multiple weights may apply
 * to a given node or edge.
 */
export type WeightsT = {|
  nodeWeightsT: NodeWeightsT,
  // Map from an edge prefix or address to a weight
  edgeWeightsT: EdgeWeightsT,
|};

/**
 * Transform NodeWeightsT and EdgeWeightsT into WeightsT.
 */
export function toWeightsT(
  nodeWeightsT: NodeWeightsT,
  edgeWeightsT: EdgeWeightsT
): WeightsT {
  return {
    nodeWeightsT,
    edgeWeightsT,
  };
}

/**
 * Creates new, empty weights.
 */
export function empty(): WeightsT {
  return {
    nodeWeightsT: emptyNodeWeightsT(),
    edgeWeightsT: emptyEdgeWeightsT(),
  };
}

export function copy(w: WeightsT): WeightsT {
  return {
    nodeWeightsT: copyNodeWeightsT(w.nodeWeightsT),
    edgeWeightsT: copyEdgeWeightsT(w.edgeWeightsT),
  };
}

/** Merge multiple Weights together.
 *
 * The resultant Weights will have every weight specified by each of the input
 * weights.
 *
 * When there are overlaps (i.e. the same address is present in two or more of
 * the Weights), then the appropriate resolver will be invoked to resolve the
 * conflict. The resolver takes two weights and combines them to return a new
 * weight.
 *
 * When no resolvers are explicitly provided, merge defaults to
 * conservative "error on conflict" resolvers.
 */
export function merge(
  ws: $ReadOnlyArray<WeightsT>,
  resolvers: ?WeightsTResolvers
): WeightsT {
  if (resolvers == null) {
    const nodeResolver = (_unused_a, _unused_b) => {
      throw new Error(
        "node weight conflict detected, but no resolver specified"
      );
    };
    const edgeResolver = (_unused_a, _unused_b) => {
      throw new Error(
        "edge weight conflict detected, but no resolver specified"
      );
    };
    resolvers = {nodeResolver, edgeResolver};
  }
  const weights: WeightsT = empty();
  const {nodeWeightsT, edgeWeightsT} = weights;
  const {nodeResolver, edgeResolver} = resolvers;
  for (const w of ws) {
    for (const [addr, val] of w.nodeWeightsT.entries()) {
      const existing = nodeWeightsT.get(addr);
      if (existing == null) {
        nodeWeightsT.set(addr, val);
      } else {
        try {
          nodeWeightsT.set(addr, nodeResolver(existing, val));
        } catch (e) {
          throw new Error(`${e} when resolving ${NodeAddress.toString(addr)}`);
        }
      }
    }
    for (const [addr, val] of w.edgeWeightsT.entries()) {
      const existing = edgeWeightsT.get(addr);
      if (existing == null) {
        edgeWeightsT.set(addr, val);
      } else {
        try {
          edgeWeightsT.set(addr, edgeResolver(existing, val));
        } catch (e) {
          throw new Error(
            `Error ${e} when resolving ${EdgeAddress.toString(addr)}`
          );
        }
      }
    }
  }
  return weights;
}

export type SerializedWeights_0_2_0 = {|
  +nodeWeightsT: {[NodeAddressT]: NodeWeight},
  +edgeWeightsT: {[EdgeAddressT]: EdgeWeight},
|};

function serialize_0_2_0(weights: WeightsT): SerializedWeights_0_2_0 {
  return {
    nodeWeightsT: MapUtil.toObject(weights.nodeWeightsT),
    edgeWeightsT: MapUtil.toObject(weights.edgeWeightsT),
  };
}

function deserialize_0_2_0(weights: SerializedWeights_0_2_0): WeightsT {
  return {
    nodeWeightsT: MapUtil.fromObject(weights.nodeWeightsT),
    edgeWeightsT: MapUtil.fromObject(weights.edgeWeightsT),
  };
}

const Parse_0_2_0: C.Parser<SerializedWeights_0_2_0> = (() => {
  const parseEdgeWeight = C.object({forwards: C.number, backwards: C.number});
  return C.object({
    nodeWeightsT: C.dict(C.number, NodeAddress.parser),
    edgeWeightsT: C.dict(parseEdgeWeight, EdgeAddress.parser),
  });
})();

const COMPAT_INFO = {type: "sourcecred/weights", version: "0.2.0"};

export const parser: C.Parser<WeightsT> = compatibleParser(COMPAT_INFO.type, {
  "0.2.0": C.fmap(Parse_0_2_0, deserialize_0_2_0),
});

export type WeightsJSON_0_2_0 = Compatible<SerializedWeights_0_2_0>;
export type WeightsJSON = WeightsJSON_0_2_0;

export function toJSON(weights: WeightsT): WeightsJSON {
  return toCompat(COMPAT_INFO, serialize_0_2_0(weights));
}

export function fromJSON(json: WeightsJSON): WeightsT {
  return parser.parseOrThrow(json);
}

export type NodeWeightDiff = {|
  +address: NodeAddressT,
  +first: ?NodeWeight,
  +second: ?NodeWeight,
|};

export type EdgeWeightDiff = {|
  +address: EdgeAddressT,
  +first: ?EdgeWeight,
  +second: ?EdgeWeight,
|};

export type WeightsComparison = {|
  +weightsAreEqual: boolean,
  +nodeWeightDiffs: $ReadOnlyArray<NodeWeightDiff>,
  +edgeWeightDiffs: $ReadOnlyArray<EdgeWeightDiff>,
|};

export function compareWeightsT(
  firstWeights: WeightsT,
  secondWeights: WeightsT
): WeightsComparison {
  const nodeWeightDiffs = [];
  const edgeWeightDiffs = [];

  const nodeAddresses = new Set([
    ...firstWeights.nodeWeightsT.keys(),
    ...secondWeights.nodeWeightsT.keys(),
  ]);
  for (const address of nodeAddresses) {
    const first = firstWeights.nodeWeightsT.get(address);
    const second = secondWeights.nodeWeightsT.get(address);
    if (!deepEqual(first, second)) {
      nodeWeightDiffs.push({
        address,
        first,
        second,
      });
    }
  }
  const edgeAddresses = new Set([
    ...firstWeights.edgeWeightsT.keys(),
    ...secondWeights.edgeWeightsT.keys(),
  ]);
  for (const address of edgeAddresses) {
    const first = firstWeights.edgeWeightsT.get(address);
    const second = secondWeights.edgeWeightsT.get(address);
    if (!deepEqual(first, second)) {
      edgeWeightDiffs.push({
        address,
        first,
        second,
      });
    }
  }
  const weightsAreEqual =
    nodeWeightDiffs.length === 0 && edgeWeightDiffs.length === 0;

  return {
    weightsAreEqual,
    nodeWeightDiffs,
    edgeWeightDiffs,
  };
}
