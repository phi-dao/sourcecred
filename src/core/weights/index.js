// @flow

import {type WeightsI, Weights} from "./weights";

export type {WeightsI};
export {Weights};

import {
  type WeightsT,
  type WeightsJSON,
  empty,
  copy,
  merge,
  toJSON,
  fromJSON,
  compareWeightsT,
} from "./weightsT";

export type {WeightsT, WeightsJSON};
export {empty, copy, merge, toJSON, fromJSON, compareWeightsT};
