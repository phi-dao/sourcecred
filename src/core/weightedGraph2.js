// @flow

import {
  type Edge,
  type EdgesOptions,
  type Node,
  type EdgeAddressT,
  type NodeAddressT,
  Graph,
} from "./graph";
import {type NodeWeight} from "./weights/nodeWeights";
import {type EdgeWeight} from "./weights/edgeWeights";
import {type WeightsComparison} from "./weights/weightsT";
import {Weights, type WeightsI} from "./weights";

/**
 * An entry in the array returned by AddressModule.toParts()
 */
export type AddressPart = string;

export type WeightedNode = {|
  +node: Node,
  +weight: NodeWeight,
|};

export type WeightedEdge = {|
  +edge: Edge,
  +weight: EdgeWeight,
|};

/**
 * The WeightedGraph2 class is a replacement for the WeightedGraph type. This class
 * eliminates the need give every node and edge that exists a weight. Rather,
 * address prefixes can be assigned a weight and when a node or edge weight is
 * queried, the product of matching weights along the address chain is returned.
 *
 * Over the longer-term, WeightedGraph should be refactored out of existence and
 * WeightedGraph2 should completely replace it.
 */
export class WeightedGraph2 {
  _weights: WeightsI;
  graph: Graph;

  constructor(graph?: Graph) {
    this.graph = graph ? graph : new Graph();
    this._weights = new Weights();
  }

  /**
   * Set the weight for a node address prefix.
   */
  setNodePrefixWeight(prefix: NodeAddressT, w: NodeWeight): this {
    this._weights.setNodeWeight(prefix, w);
    return this;
  }

  /**
   * Set the weights for an edge address prefix.
   */
  setEdgePrefixWeight(prefix: EdgeAddressT, w: EdgeWeight): this {
    this._weights.setEdgeWeight(prefix, w);
    return this;
  }

  /**
   * Add a node and optionally assign its weight. If a weight is not passed in,
   * the weight will default to 1.
   */
  addNode(n: Node, w?: NodeWeight): this {
    this.graph.addNode(n);
    if (w != null) this._weights.setNodeWeight(n.address, w);
    return this;
  }

  node(n: NodeAddressT): ?WeightedNode {
    const node = this.graph.node(n);
    if (node) {
      return {
        node,
        weight: this._weights.getNodeWeight(n),
      };
    }
  }

  nodes(options?: {|+prefix: NodeAddressT|}): Iterator<WeightedNode> {
    return this._nodesIterator(options);
  }

  *_nodesIterator(options?: {|+prefix: NodeAddressT|}): Iterator<WeightedNode> {
    const nodes = this.graph.nodes(options);
    for (const node of nodes) {
      const weight = this._weights.getNodeWeight(node.address);
      yield {node, weight};
    }
  }

  /**
   * Add an edge and optionally assign its weight. If a prefix of the edge's
   * address doesn't have a weight assigned, the weight will default to 1.
   */
  addEdge(e: Edge, w?: EdgeWeight): this {
    this.graph.addEdge(e);
    if (w != null) this._weights.setEdgeWeight(e.address, w);
    return this;
  }

  edge(e: EdgeAddressT): ?WeightedEdge {
    const edge = this.graph.edge(e);
    if (edge) {
      return {
        edge,
        weight: this._weights.getEdgeWeight(e),
      };
    }
  }

  edges(options: EdgesOptions): Iterator<WeightedEdge> {
    return this._edgesIterator(options);
  }

  *_edgesIterator(options: EdgesOptions): Iterator<WeightedEdge> {
    const edges = this.graph.edges(options);
    for (const edge of edges) {
      const weight = this._weights.getEdgeWeight(edge.address);
      yield {edge, weight};
    }
  }

  /**
   * Return the weight of a Node address.
   */
  getNodeAddressWeight(_unused_a: NodeAddressT): NodeWeight {
    throw new Error("method not implemented");
  }

  /**
   * Return the weight of an Edge address.
   */
  getEdgeAddressWeight(_unused_a: EdgeAddressT): EdgeWeight {
    throw new Error("method not implemented");
  }

  // Utilities

  merge(_unused_wg: WeightedGraph2): WeightedGraph2 {
    throw new Error("method not implemented");
  }

  compareWeights(_unused_wg: WeightedGraph2): WeightsComparison {
    throw new Error("method not implemented");
  }

  // TODO: add JSON serialization and deserialization helpers
  // TODO: Add a tagging system so nodes and edges can have their weights
  // programmatically shifted
}
