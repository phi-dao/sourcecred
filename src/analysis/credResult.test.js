// @flow

import stringify from "json-stable-stringify";
import {Graph, NodeAddress, EdgeAddress} from "../core/graph";
import * as WeightsT from "../core/weights/weightsT";
import type {NodeType, EdgeType} from "./types";
import type {PluginDeclaration} from "./pluginDeclaration";
import {defaultParams} from "./timeline/params";
import {compute, toJSON, fromJSON} from "./credResult";
import {IDENTITY_PREFIX} from "../core/identity";

describe("analysis/credResult", () => {
  describe("to/fro JSON", () => {
    async function example() {
      const nodeType: NodeType = {
        name: "node",
        pluralName: "nodes",
        prefix: IDENTITY_PREFIX,
        defaultWeight: 2,
        description: "a type",
      };
      const edgeType: EdgeType = {
        forwardName: "points",
        backwardName: "is pointed to",
        prefix: EdgeAddress.fromParts(["edge"]),
        defaultWeight: {forwards: 2, backwards: 3},
        description: "a type",
      };
      const declaration: PluginDeclaration = {
        name: "empty",
        nodePrefix: NodeAddress.empty,
        edgePrefix: EdgeAddress.empty,
        nodeTypes: [nodeType],
        edgeTypes: [edgeType],
        userTypes: [nodeType],
      };

      const graph = new Graph()
        .addNode({
          address: NodeAddress.append(IDENTITY_PREFIX, "1"),
          description: "n1",
          timestampMs: 100000,
        })
        .addNode({
          address: NodeAddress.append(IDENTITY_PREFIX, "2"),
          description: "n2",
          timestampMs: 110000,
        });
      const weights = WeightsT.empty();
      const wg = {graph, weights};
      const params = defaultParams();
      const result = await compute(wg, params, [declaration], []);
      return result;
    }

    describe("to/froJSON", () => {
      it("round-trips appropriately", async () => {
        const r = await example();
        const j = toJSON(r);
        const r_ = fromJSON(j);
        const j_ = toJSON(r_);

        expect(r.weightedGraph.graph.equals(r_.weightedGraph.graph)).toBe(true);
        expect(r.weightedGraph.weights).toEqual(r_.weightedGraph.weights);
        expect(r.credData).toEqual(r_.credData);
        expect(r.params).toEqual(r_.params);
        expect(r.plugins).toEqual(r_.plugins);
        expect(j).toEqual(j_);
      });
      it("snapshots as expected", async () => {
        const r = await example();
        const j = toJSON(r);
        expect(stringify(j, {space: 2})).toMatchSnapshot();
      });
    });
  });
});
