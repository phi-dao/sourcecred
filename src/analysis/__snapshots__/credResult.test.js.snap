// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analysis/credResult to/fro JSON to/froJSON snapshots as expected 1`] = `
"[
  {
    \\"type\\": \\"sourcecred/credResult\\",
    \\"version\\": \\"0.1.0\\"
  },
  {
    \\"credData\\": {
      \\"edgeOverTime\\": [
      ],
      \\"edgeSummaries\\": [
      ],
      \\"intervals\\": [
        {
          \\"endTimeMs\\": 259200000,
          \\"startTimeMs\\": -345600000
        }
      ],
      \\"nodeOverTime\\": [
        {
          \\"cred\\": [
            0.5
          ],
          \\"dependencyMintedCred\\": [
            0
          ],
          \\"seedFlow\\": [
            0.1
          ],
          \\"syntheticLoopFlow\\": [
            0.4
          ]
        },
        {
          \\"cred\\": [
            0.5
          ],
          \\"dependencyMintedCred\\": [
            0
          ],
          \\"seedFlow\\": [
            0.1
          ],
          \\"syntheticLoopFlow\\": [
            0.4
          ]
        }
      ],
      \\"nodeSummaries\\": [
        {
          \\"cred\\": 0.5,
          \\"dependencyMintedCred\\": 0,
          \\"seedFlow\\": 0.1,
          \\"syntheticLoopFlow\\": 0.4
        },
        {
          \\"cred\\": 0.5,
          \\"dependencyMintedCred\\": 0,
          \\"seedFlow\\": 0.1,
          \\"syntheticLoopFlow\\": 0.4
        }
      ]
    },
    \\"dependencyPolicies\\": [
    ],
    \\"params\\": {
      \\"alpha\\": 0.2,
      \\"intervalDecay\\": 0.5
    },
    \\"plugins\\": [
      {
        \\"type\\": \\"sourcecred/pluginDeclarations\\",
        \\"version\\": \\"0.1.0\\"
      },
      [
        {
          \\"edgePrefix\\": \\"E\\\\u0000\\",
          \\"edgeTypes\\": [
            {
              \\"backwardName\\": \\"is pointed to\\",
              \\"defaultWeight\\": {
                \\"backwards\\": 3,
                \\"forwards\\": 2
              },
              \\"description\\": \\"a type\\",
              \\"forwardName\\": \\"points\\",
              \\"prefix\\": \\"E\\\\u0000edge\\\\u0000\\"
            }
          ],
          \\"name\\": \\"empty\\",
          \\"nodePrefix\\": \\"N\\\\u0000\\",
          \\"nodeTypes\\": [
            {
              \\"defaultWeight\\": 2,
              \\"description\\": \\"a type\\",
              \\"name\\": \\"node\\",
              \\"pluralName\\": \\"nodes\\",
              \\"prefix\\": \\"N\\\\u0000sourcecred\\\\u0000core\\\\u0000IDENTITY\\\\u0000\\"
            }
          ],
          \\"userTypes\\": [
            {
              \\"defaultWeight\\": 2,
              \\"description\\": \\"a type\\",
              \\"name\\": \\"node\\",
              \\"pluralName\\": \\"nodes\\",
              \\"prefix\\": \\"N\\\\u0000sourcecred\\\\u0000core\\\\u0000IDENTITY\\\\u0000\\"
            }
          ]
        }
      ]
    ],
    \\"weightedGraph\\": [
      {
        \\"type\\": \\"sourcecred/weightedGraph\\",
        \\"version\\": \\"0.1.0\\"
      },
      {
        \\"graphJSON\\": [
          {
            \\"type\\": \\"sourcecred/graph\\",
            \\"version\\": \\"0.8.0\\"
          },
          {
            \\"edges\\": [
            ],
            \\"nodes\\": [
              {
                \\"description\\": \\"n1\\",
                \\"index\\": 0,
                \\"timestampMs\\": 100000
              },
              {
                \\"description\\": \\"n2\\",
                \\"index\\": 1,
                \\"timestampMs\\": 110000
              }
            ],
            \\"sortedNodeAddresses\\": [
              [
                \\"sourcecred\\",
                \\"core\\",
                \\"IDENTITY\\",
                \\"1\\"
              ],
              [
                \\"sourcecred\\",
                \\"core\\",
                \\"IDENTITY\\",
                \\"2\\"
              ]
            ]
          }
        ],
        \\"weightsJSON\\": [
          {
            \\"type\\": \\"sourcecred/weights\\",
            \\"version\\": \\"0.2.0\\"
          },
          {
            \\"edgeWeightsT\\": {
            },
            \\"nodeWeightsT\\": {
            }
          }
        ]
      }
    ]
  }
]"
`;
